iree/hal/metal/metal_command_buffer.mm

```diff
@@ -215,20 +249,125 @@
+      const PipelineStateObject::PushState& push_state = push_kv.second;
+      IREE_DVLOG(3) << " For set #" << set_number;
+
+      id<MTLArgumentEncoder> argument_encoder =
+          [metal_kernel newArgumentEncoderWithBufferIndex:set_number];  // retained
+      if (!argument_encoder) {
+        return InvalidArgumentErrorBuilder(IREE_LOC)
+               << "Buffer index #" << set_number << " is not an argument buffer";
+      }
+
+      __block id<MTLBuffer> argument_buffer =
+          [metal_handle_.device newBufferWithLength:argument_encoder.encodedLength
+                                            options:MTLResourceStorageModeShared];  // retained
+      if (!argument_buffer) {
+        return InternalErrorBuilder(IREE_LOC)
+               << "Failed to create argument buffer with length=" << argument_encoder.encodedLength;
+      }
+      [metal_handle_ addCompletedHandler:^(id<MTLCommandBuffer>) {
+        [argument_buffer release];
+        [argument_encoder release];
+      }];
+
+      [argument_encoder setArgumentBuffer:argument_buffer offset:0];
+
+      for (const auto& resource_binding : push_state.resource_bindings) {
+        IREE_DVLOG(3) << "  Resource @[" << resource_binding.DebugStringShort() << "]";
+
+        if (resource_binding.length != kWholeBuffer &&
+            resource_binding.length != resource_binding.buffer->allocation_size()) {
+          return UnimplementedErrorBuilder(IREE_LOC)
+                 << "MetalCommandBuffer::Dispatch with sub-buffer";
+        }
+
+        IREE_ASSIGN_OR_RETURN(auto buffer, CastBuffer(resource_binding.buffer));
+        [argument_encoder setBuffer:buffer->handle()
+                             offset:resource_binding.offset
+                            atIndex:resource_binding.binding];
+
+        const auto* set_layout = pipeline_layout->set_layouts()[set_number];
+        const auto* layout_binding = set_layout->GetBindingForIndex(resource_binding.binding);
+        if (!layout_binding) {
+          return InvalidArgumentErrorBuilder(IREE_LOC)
+                 << "Cannot find binding #" << resource_binding.binding
+                 << " in argument buffer layout";
+        }
+        [compute_encoder useResource:buffer->handle()
+                               usage:ConvertResourceUsage(layout_binding->access)];
+      }
+
+      [compute_encoder setBuffer:argument_buffer offset:0 atIndex:set_number];
+    }
+
+    if (!pso.bind_states.empty()) {
+      return UnimplementedErrorBuilder(IREE_LOC)
+             << "MetalCommandBuffer::Dispatch with bound descriptor sets";
+    }
+
+    if (!pso.constant_states.empty()) {
+      return UnimplementedErrorBuilder(IREE_LOC)
+             << "MetalCommandBuffer::Dispatch with push constants";
+    }
+  }
+
+  IREE_DVLOG(2) << "Dispatch workgroup count: (" << workgroups[0] << ", " << workgroups[1] << ", "
+                << workgroups[2] << "), workgroup size: (32, 1, 1)";
+  // TODO(antiagainst): fix workgroup size
```

Comment:
This will happen in a future pull request for https://github.com/antiagainst/iree/commit/1e602944cfacf5ae1cc1cc84e48d93823b10ba22.